File: src/week5/P003.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
let lines = [];
rl.on('line', (line) => {
  lines.push(line);
});
rl.on('close', () => {
  if (lines.length === 0) return;

  const [v, e] = lines[0].split(' ').map(Number);
  let edges = [];
  for (let i = 1; i <= e; i++) {
    edges.push(lines[i].split(' ').map(Number)); // [a, b, c]
  }

  let parent = Array.from({ length: v + 1 }, (_, i) => i);
  let res = 0;

  edges.sort((a, b) => a[2] - b[2]);

  // Union-Find 알고리즘
  function find(x) {
    if (x === parent[x]) {
      return x;
    }
    return (parent[x] = find(parent[x]));
  }

  function union(x, y) {
    x = find(x);
    y = find(y);

    if (x !== y) {
      if (x <= y) {
        parent[y] = x;
      } else {
        parent[x] = y;
      }
    }
  }
  // 크루스칼 알고리즘
  for (let i = 0; i < e; i++) {
    const [x, y, c] = edges[i];
    if (find(x) !== find(y)) {
      // 부모 노드가 다름 (사이클 형성 X)
      union(x, y); // 최소 신장 트리에 포함시킴
      res += c;
    }
  }
  console.log(res);
});

File: src/week5/P003.java
package week5;

import java.io.*;
import java.util.*;

public class P003 {
    static int[] parent;

    // 간선 정보를 저장할 클래스 (정렬을 위해 Comparable 구현)
    static class Edge implements Comparable<Edge> {
        int u, v, c;

        Edge(int u, int v, int c) {
            this.u = u;
            this.v = v;
            this.c = c;
        }

        @Override
        public int compareTo(Edge o) {
            return this.c - o.c; // 비용 기준 오름차순 정렬
        }
    }

    // Union-Find 알고리즘
    static int find(int x) {
        if (x == parent[x]) {
            return x;
        }
        // 경로 압축 (Path Compression)
        return parent[x] = find(parent[x]);
    }

    static void union(int x, int y) {
        x = find(x);
        y = find(y);

        if (x != y) {
            if (x <= y) {
                parent[y] = x;
            } else {
                parent[x] = y;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            edges.add(new Edge(a, b, c));
        }

        parent = new int[v + 1];
        for (int i = 1; i <= v; i++) {
            parent[i] = i;
        }

        long res = 0; // 결과값 (비용 합)

        // 간선을 최소 비용 순으로 오름차순 정렬
        Collections.sort(edges);

        // 크루스칼 알고리즘
        for (Edge edge : edges) {
            if (find(edge.u) != find(edge.v)) { // 부모 노드가 다름
                union(edge.u, edge.v); // 최소 신장 트리에 포함시킴
                res += edge.c;
            }
        }

        System.out.println(res);
    }
}
